// expo-nexen-player/lib/module/hoc/withAnimation.js
function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

import React from 'react';
import { Animated } from 'react-native';
const ANIMATION_DURATION = 350;
const USE_NATIVE_DRIVER = false;
export const withAnimation = ComposedComponent => {
  const WrapperComponent = /*#__PURE__*/React.forwardRef((props, ref) => {
    const {
      animateFrom,
      distance,
      style,
      ...composedComponentProps
    } = props;
    const opacity = React.useRef(new Animated.Value(0)).current;
    const margin = React.useRef(new Animated.Value(-distance)).current;

    const startComponentShowAnimation = callback => {
      Animated.parallel([Animated.timing(opacity, {
        toValue: 1,
        duration: ANIMATION_DURATION,
        useNativeDriver: USE_NATIVE_DRIVER
      }), Animated.timing(margin, {
        toValue: 0,
        duration: ANIMATION_DURATION,
        useNativeDriver: USE_NATIVE_DRIVER
      })]).start(callback);
    };

    const startComponentHideAnimation = callback => {
      Animated.parallel([Animated.timing(opacity, {
        toValue: 0,
        duration: ANIMATION_DURATION,
        useNativeDriver: USE_NATIVE_DRIVER
      }), Animated.timing(margin, {
        toValue: -distance,
        duration: ANIMATION_DURATION,
        useNativeDriver: USE_NATIVE_DRIVER
      })]).start(callback);
    };

    React.useImperativeHandle(ref, () => ({
      show: callback => {
        startComponentShowAnimation(callback);
      },
      hide: callback => {
        startComponentHideAnimation(callback);
      }
    }));
    const wrapperStyle = {
      opacity,
      marginLeft: animateFrom === 'LEFT' ? margin : undefined,
      marginTop: animateFrom === 'TOP' ? margin : undefined,
      marginRight: animateFrom === 'RIGHT' ? margin : undefined,
      marginBottom: animateFrom === 'BOTTOM' ? margin : undefined
    };
    return /*#__PURE__*/React.createElement(ComposedComponent, _extends({
      style: [style, wrapperStyle]
    }, composedComponentProps));
  });
  return WrapperComponent;
};
