// expo-nexen-player/lib/module/components/ErrorView.js
import React from 'react';
import { StyleSheet, Text } from 'react-native';
import { IconXCircle } from '../assets/icons';
import ModalView from './ModalView';

const ErrorView = props => {
  const {
    errorText,
    theme,
    style
  } = props;
  const errorTextStyle = {
    color: theme === null || theme === void 0 ? void 0 : theme.textColor,
    fontSize: theme === null || theme === void 0 ? void 0 : theme.textSize,
    fontFamily: theme === null || theme === void 0 ? void 0 : theme.font
  };
  return /*#__PURE__*/React.createElement(ModalView, {
    style: [style]
  }, /*#__PURE__*/React.createElement(IconXCircle, {
    size: theme === null || theme === void 0 ? void 0 : theme.iconSize,
    color: theme === null || theme === void 0 ? void 0 : theme.iconColor
  }), /*#__PURE__*/React.createElement(Text, {
    style: [styles.text, errorTextStyle]
  }, errorText));
};

export default ErrorView;
const styles = StyleSheet.create({
  container: {
    justifyContent: 'center',
    alignItems: 'center'
  },
  text: {
    marginTop: 5
  }
});
