// expo-nexen-player/lib/module/components/GradientView.js
import React from 'react';
import { StyleSheet, View } from 'react-native';
import Svg, { Defs, LinearGradient, Rect, Stop } from 'react-native-svg';

const GradientView = props => {
  const {
    startPoint,
    endPoint,
    startColor,
    middleColor,
    endColor,
    startOpacity,
    middleOpacity,
    endOpacity,
    startOffset,
    middleOffset,
    endOffset,
    style
  } = props;
  const [dimensions, setDimensions] = React.useState({
    width: 0,
    height: 0
  });

  const onLayoutChange = async e => {
    const {
      width,
      height
    } = e.nativeEvent.layout;
    const {
      width: w,
      height: h
    } = dimensions;

    if (w !== width || h !== height) {
      setDimensions({
        width,
        height
      });
    }
  };

  return /*#__PURE__*/React.createElement(View, {
    style: [styles.container, style],
    onLayout: onLayoutChange
  }, /*#__PURE__*/React.createElement(Svg, {
    width: dimensions.width,
    height: dimensions.height
  }, /*#__PURE__*/React.createElement(Defs, null, /*#__PURE__*/React.createElement(LinearGradient, {
    id: "gradient",
    x1: startPoint === null || startPoint === void 0 ? void 0 : startPoint.x,
    y1: startPoint === null || startPoint === void 0 ? void 0 : startPoint.y,
    x2: endPoint === null || endPoint === void 0 ? void 0 : endPoint.x,
    y2: endPoint === null || endPoint === void 0 ? void 0 : endPoint.y
  }, /*#__PURE__*/React.createElement(Stop, {
    offset: startOffset,
    stopColor: startColor,
    stopOpacity: startOpacity
  }), /*#__PURE__*/React.createElement(Stop, {
    offset: middleOffset,
    stopColor: middleColor,
    stopOpacity: middleOpacity
  }), /*#__PURE__*/React.createElement(Stop, {
    offset: endOffset,
    stopColor: endColor,
    stopOpacity: endOpacity
  }))), /*#__PURE__*/React.createElement(Rect, {
    x: "0",
    y: "0",
    width: dimensions.width,
    height: dimensions.height,
    fill: "url(#gradient)"
  })));
};

export default GradientView;
GradientView.defaultProps = {
  startPoint: {
    x: 0,
    y: 0
  },
  endPoint: {
    x: 0,
    y: 1
  },
  startColor: 'black',
  middleColor: 'black',
  endColor: 'black',
  startOpacity: 1,
  middleOpacity: 1,
  endOpacity: 1,
  startOffset: 0,
  middleOffset: 0.5,
  endOffset: 1
};
const styles = StyleSheet.create({
  container: {
    width: '100%',
    height: '25%'
  }
});
