// expo-nexen-player/lib/module/components/SpeedControl.js
import React from 'react';
import { StyleSheet, TouchableHighlight } from 'react-native';
import { getAlphaColor } from '../utils/ColorUtil';
import GradientView from './GradientView';
import StepSeekBar from './StepSeekBar';
import ModalView from './ModalView';
import { withAnimation } from '../hoc/withAnimation';
const speedData = ['0.25', '0.5', '0.75', '1.0', '1.5', '2.0', '3.0'];

const SpeedControl = props => {
  const {
    style,
    currentSpeed,
    fullScreen,
    nexenTheme,
    insets,
    onSpeedChange
  } = props;
  const curerntIndex = speedData.indexOf(currentSpeed);
  const validIndex = curerntIndex < 0 ? 3 : curerntIndex;
  const CONTAINER_HORIZONTAL_PADDING = fullScreen ? ((insets === null || insets === void 0 ? void 0 : insets.left) + (insets === null || insets === void 0 ? void 0 : insets.right)) / 2 > 0 ? ((insets === null || insets === void 0 ? void 0 : insets.left) + (insets === null || insets === void 0 ? void 0 : insets.right)) / 2 : 8 : 8;
  const stepSeekBarTheme = React.useMemo(() => {
    var _nexenTheme$fonts, _nexenTheme$speedSeek, _nexenTheme$colors, _nexenTheme$speedSeek2, _nexenTheme$colors2, _nexenTheme$speedSeek3, _nexenTheme$colors3, _nexenTheme$speedSeek4, _nexenTheme$colors4, _nexenTheme$speedSeek5, _nexenTheme$speedSeek6, _nexenTheme$speedSeek7, _nexenTheme$speedSeek8, _nexenTheme$speedSeek9, _nexenTheme$speedSeek10;

    return {
      font: nexenTheme === null || nexenTheme === void 0 ? void 0 : (_nexenTheme$fonts = nexenTheme.fonts) === null || _nexenTheme$fonts === void 0 ? void 0 : _nexenTheme$fonts.secondaryFont,
      lineColor: (nexenTheme === null || nexenTheme === void 0 ? void 0 : (_nexenTheme$speedSeek = nexenTheme.speedSeekBar) === null || _nexenTheme$speedSeek === void 0 ? void 0 : _nexenTheme$speedSeek.lineColor) || getAlphaColor(nexenTheme === null || nexenTheme === void 0 ? void 0 : (_nexenTheme$colors = nexenTheme.colors) === null || _nexenTheme$colors === void 0 ? void 0 : _nexenTheme$colors.accentColor, 0.7),
      dotColor: (nexenTheme === null || nexenTheme === void 0 ? void 0 : (_nexenTheme$speedSeek2 = nexenTheme.speedSeekBar) === null || _nexenTheme$speedSeek2 === void 0 ? void 0 : _nexenTheme$speedSeek2.dotColor) || (nexenTheme === null || nexenTheme === void 0 ? void 0 : (_nexenTheme$colors2 = nexenTheme.colors) === null || _nexenTheme$colors2 === void 0 ? void 0 : _nexenTheme$colors2.secondaryColor),
      thumbColor: (nexenTheme === null || nexenTheme === void 0 ? void 0 : (_nexenTheme$speedSeek3 = nexenTheme.speedSeekBar) === null || _nexenTheme$speedSeek3 === void 0 ? void 0 : _nexenTheme$speedSeek3.thumbColor) || (nexenTheme === null || nexenTheme === void 0 ? void 0 : (_nexenTheme$colors3 = nexenTheme.colors) === null || _nexenTheme$colors3 === void 0 ? void 0 : _nexenTheme$colors3.secondaryColor),
      textColor: (nexenTheme === null || nexenTheme === void 0 ? void 0 : (_nexenTheme$speedSeek4 = nexenTheme.speedSeekBar) === null || _nexenTheme$speedSeek4 === void 0 ? void 0 : _nexenTheme$speedSeek4.textColor) || (nexenTheme === null || nexenTheme === void 0 ? void 0 : (_nexenTheme$colors4 = nexenTheme.colors) === null || _nexenTheme$colors4 === void 0 ? void 0 : _nexenTheme$colors4.secondaryTextColor),
      textSize: nexenTheme === null || nexenTheme === void 0 ? void 0 : (_nexenTheme$speedSeek5 = nexenTheme.speedSeekBar) === null || _nexenTheme$speedSeek5 === void 0 ? void 0 : _nexenTheme$speedSeek5.textSize,
      lineHeight: nexenTheme === null || nexenTheme === void 0 ? void 0 : (_nexenTheme$speedSeek6 = nexenTheme.speedSeekBar) === null || _nexenTheme$speedSeek6 === void 0 ? void 0 : _nexenTheme$speedSeek6.lineHeight,
      dotSize: nexenTheme === null || nexenTheme === void 0 ? void 0 : (_nexenTheme$speedSeek7 = nexenTheme.speedSeekBar) === null || _nexenTheme$speedSeek7 === void 0 ? void 0 : _nexenTheme$speedSeek7.dotSize,
      thumbSize: nexenTheme === null || nexenTheme === void 0 ? void 0 : (_nexenTheme$speedSeek8 = nexenTheme.speedSeekBar) === null || _nexenTheme$speedSeek8 === void 0 ? void 0 : _nexenTheme$speedSeek8.thumbSize,
      dotCornerRadius: nexenTheme === null || nexenTheme === void 0 ? void 0 : (_nexenTheme$speedSeek9 = nexenTheme.speedSeekBar) === null || _nexenTheme$speedSeek9 === void 0 ? void 0 : _nexenTheme$speedSeek9.dotCornerRadius,
      thumbCornerRadius: nexenTheme === null || nexenTheme === void 0 ? void 0 : (_nexenTheme$speedSeek10 = nexenTheme.speedSeekBar) === null || _nexenTheme$speedSeek10 === void 0 ? void 0 : _nexenTheme$speedSeek10.thumbCornerRadius
    };
  }, [nexenTheme]);
  const SEEK_BAR_HEIGHT = stepSeekBarTheme.thumbSize + 12 * 2 + (stepSeekBarTheme === null || stepSeekBarTheme === void 0 ? void 0 : stepSeekBarTheme.textSize);
  const containerStyle = {
    left: CONTAINER_HORIZONTAL_PADDING,
    right: CONTAINER_HORIZONTAL_PADDING,
    bottom: 0,
    paddingHorizontal: 0,
    paddingVertical: 0
  };
  const stepSeekBarStyle = {
    height: SEEK_BAR_HEIGHT
  };
  return /*#__PURE__*/React.createElement(ModalView, {
    style: [styles.container, style, containerStyle]
  }, /*#__PURE__*/React.createElement(GradientView, {
    style: {
      height: '100%',
      width: '100%'
    },
    startOpacity: 0.0,
    middleOpacity: 0.2,
    endOpacity: 0.5
  }), /*#__PURE__*/React.createElement(TouchableHighlight, {
    style: [styles.innerContainer]
  }, /*#__PURE__*/React.createElement(StepSeekBar, {
    style: stepSeekBarStyle,
    data: speedData,
    initialIndex: validIndex,
    theme: stepSeekBarTheme,
    onStepChange: onSpeedChange
  })));
};

export default withAnimation(SpeedControl);
SpeedControl.defaultProps = {};
const styles = StyleSheet.create({
  container: {
    maxHeight: 120,
    zIndex: 110,
    overflow: 'hidden'
  },
  innerContainer: {
    position: 'absolute',
    left: 0,
    top: 0,
    right: 0,
    bottom: 0,
    padding: 16,
    justifyContent: 'center'
  }
});
