// expo-nexen-player/lib/module/components/TimeTagView.js
import React from 'react';
import { StyleSheet, Text, View } from 'react-native';

const TimeTagView = props => {
  const {
    timeText,
    theme,
    style
  } = props;
  const textContainerStyle = {
    height: theme === null || theme === void 0 ? void 0 : theme.height,
    borderRadius: theme === null || theme === void 0 ? void 0 : theme.cornerRadius,
    borderWidth: theme === null || theme === void 0 ? void 0 : theme.borderWidth,
    borderColor: theme === null || theme === void 0 ? void 0 : theme.borderColor,
    backgroundColor: theme === null || theme === void 0 ? void 0 : theme.backgroundColor
  };
  const timeTextStyle = {
    fontSize: theme === null || theme === void 0 ? void 0 : theme.textSize,
    lineHeight: theme === null || theme === void 0 ? void 0 : theme.textLineHeight,
    color: theme === null || theme === void 0 ? void 0 : theme.textColor,
    fontFamily: theme === null || theme === void 0 ? void 0 : theme.font
  };
  return /*#__PURE__*/React.createElement(View, {
    style: [styles.container, style, textContainerStyle]
  }, /*#__PURE__*/React.createElement(Text, {
    style: [styles.text, timeTextStyle]
  }, timeText));
};

export default TimeTagView;
const styles = StyleSheet.create({
  container: {
    minWidth: 40,
    height: 18,
    borderRadius: 9,
    borderWidth: 1,
    justifyContent: 'center',
    alignItems: 'center'
  },
  text: {
    fontSize: 11,
    marginHorizontal: 6,
    lineHeight: 13,
    opacity: 0.8
  }
});
