// expo-nexen-player/lib/module/components/VolumeTagView.js
import React from 'react';
import { StyleSheet, View } from 'react-native';
import SeekBar from './SeekBar';
import IconButton from './IconButton';
import { getVolumeIcon } from '../utils/ComponentUtil';
import { IconVolume } from '../assets/icons';

const VolumeTagView = props => {
  const {
    volume: playerVolume,
    totalVolume,
    muted,
    isSeekable,
    theme,
    style,
    onVolumeSeekStart,
    onVolumeSeekUpdate,
    onVolumeSeekEnd,
    onVolumePress
  } = props;
  const [volume, setVolume] = React.useState(0);
  const SEEK_WIDTH = 100;
  const seekBarTheme = React.useMemo(() => {
    return {
      trackColor: theme === null || theme === void 0 ? void 0 : theme.barColor,
      totalTrackColor: theme === null || theme === void 0 ? void 0 : theme.underlayColor,
      thumbColor: theme === null || theme === void 0 ? void 0 : theme.thumbColor,
      cachedTrackColor: 'rgba(0,0,0,0)',
      thumbBorderColor: 'rgba(0,0,0,0)',
      thumbBorderWidth: 0,
      thumbBorderCornerRadius: 0,
      trackHeight: theme === null || theme === void 0 ? void 0 : theme.trackHeight,
      thumbSize: theme === null || theme === void 0 ? void 0 : theme.thumbSize,
      thumbCornerRadius: theme === null || theme === void 0 ? void 0 : theme.thumbCornerRadius
    };
  }, [theme]);
  const containerStyle = {
    width: SEEK_WIDTH + ((theme === null || theme === void 0 ? void 0 : theme.iconSize) + 20)
  };
  const seekStyle = {
    width: SEEK_WIDTH
  };
  const onSeekStart = React.useCallback((value, totalValue, position) => {
    setVolume(value);
    onVolumeSeekStart === null || onVolumeSeekStart === void 0 ? void 0 : onVolumeSeekStart(value, totalValue, position);
  }, []);
  const onSeekUpdate = React.useCallback((value, totalValue, position) => {
    setVolume(value);
    onVolumeSeekUpdate === null || onVolumeSeekUpdate === void 0 ? void 0 : onVolumeSeekUpdate(value, totalValue, position);
  }, []);
  const onSeekEnd = React.useCallback((value, totalValue, position) => {
    setVolume(value);
    onVolumeSeekEnd === null || onVolumeSeekEnd === void 0 ? void 0 : onVolumeSeekEnd(value, totalValue, position);
  }, []);

  const onToggleVolume = () => {
    onVolumePress === null || onVolumePress === void 0 ? void 0 : onVolumePress();
  };

  React.useEffect(() => {
    setVolume(playerVolume);
  }, [playerVolume]);
  return /*#__PURE__*/React.createElement(View, {
    style: [styles.container, style, containerStyle]
  }, /*#__PURE__*/React.createElement(SeekBar, {
    style: [styles.seek, seekStyle],
    trackValue: volume,
    totalTrackValue: totalVolume,
    disableCachedTrack: true,
    disableThumbBorder: true,
    theme: seekBarTheme,
    isSeekable: isSeekable,
    onSeekStart: onSeekStart,
    onSeekUpdate: onSeekUpdate,
    onSeekEnd: onSeekEnd
  }), /*#__PURE__*/React.createElement(IconButton, {
    onPress: onToggleVolume
  }, muted ? /*#__PURE__*/React.createElement(IconVolume, {
    size: theme === null || theme === void 0 ? void 0 : theme.iconSize,
    color: theme === null || theme === void 0 ? void 0 : theme.iconColor
  }) : getVolumeIcon(volume, totalVolume, theme === null || theme === void 0 ? void 0 : theme.iconSize, theme === null || theme === void 0 ? void 0 : theme.iconColor)));
};

export default VolumeTagView;
VolumeTagView.defaultProps = {
  volume: 50,
  totalVolume: 100,
  fullScreen: false
};
const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center'
  },
  seek: {
    paddingVertical: 3
  }
});
