// expo-nexen-player/lib/commonjs/utils/Theme.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DefaultVolumeSeekBarTheme = exports.DefaultTrackSeekBarTheme = exports.DefaultTipViewTheme = exports.DefaultTheme = exports.DefaultTagViewTheme = exports.DefaultSpeedSeekBar = exports.DefaultSizesTheme = exports.DefaultMiniSeekBarTheme = exports.DefaultLockButtonTheme = exports.DefaultColorsTheme = exports.DefaultBrightnessSeekBarTheme = void 0;
const DefaultColorsTheme = {
  primaryColor: '#0a0a0a',
  secondaryColor: '#fafafa',
  accentColor: '#fa5005',
  primaryTextColor: '#fafafa',
  secondaryTextColor: '#f2f2f2',
  primaryIconColor: '#fafafa',
  secondaryIconColor: '#f2f2f2',
  rippleIconColor: '#f2f2f2'
};
exports.DefaultColorsTheme = DefaultColorsTheme;
const DefaultSizesTheme = {
  modalCornerRadius: 12,
  primaryTextSize: 20,
  secondaryTextSize: 15,
  primaryIconSize: 20,
  secondaryIconSize: 20,
  rippleIconSize: 28,
  paddingVertical: 8,
  paddingHorizontal: 4
};
exports.DefaultSizesTheme = DefaultSizesTheme;
const DefaultTrackSeekBarTheme = {
  trackHeight: 2,
  thumbSize: 10,
  thumbBorderWidth: 5,
  thumbCornerRadius: 5,
  thumbBorderCornerRadius: 7.5
};
exports.DefaultTrackSeekBarTheme = DefaultTrackSeekBarTheme;
const DefaultSpeedSeekBar = {
  textSize: 12,
  lineHeight: 2,
  dotSize: 6,
  dotCornerRadius: 3,
  thumbSize: 18,
  thumbCornerRadius: 9
};
exports.DefaultSpeedSeekBar = DefaultSpeedSeekBar;
const DefaultVolumeSeekBarTheme = {
  barColor: 'rgba(250,250,250,0.5)',
  underlayColor: 'rgba(10,10,10,0.3)'
};
exports.DefaultVolumeSeekBarTheme = DefaultVolumeSeekBarTheme;
const DefaultBrightnessSeekBarTheme = {
  barColor: 'rgba(250,250,250,0.5)',
  underlayColor: 'rgba(10,10,10,0.3)'
};
exports.DefaultBrightnessSeekBarTheme = DefaultBrightnessSeekBarTheme;
const DefaultMiniSeekBarTheme = {
  barColor: 'rgba(250,80,5,0.6)',
  underlayColor: 'rgba(250,250,250,0.3)',
  thumbColor: '#fafafa',
  trackHeight: 2,
  thumbSize: 10,
  thumbCornerRadius: 5
};
exports.DefaultMiniSeekBarTheme = DefaultMiniSeekBarTheme;
;
const DefaultLockButtonTheme = {
  textSize: 14,
  thumbIconSize: 18,
  thumbHeight: 36,
  thumbWidth: 36,
  thumbCornerRadius: 18
};
exports.DefaultLockButtonTheme = DefaultLockButtonTheme;
const DefaultTagViewTheme = {
  height: 18,
  borderWidth: 1,
  cornerRadius: 9,
  textSize: 12,
  textLineHeight: 14,
  iconSize: 12
};
exports.DefaultTagViewTheme = DefaultTagViewTheme;
const DefaultTipViewTheme = {
  iconSize: 20,
  textSize: 18
};
exports.DefaultTipViewTheme = DefaultTipViewTheme;
const DefaultTheme = {
  trackSeekBar: { ...DefaultTrackSeekBarTheme
  },
  speedSeekBar: { ...DefaultSpeedSeekBar
  },
  miniSeekBar: { ...DefaultMiniSeekBarTheme
  },
  lockButton: { ...DefaultLockButtonTheme
  },
  tipView: { ...DefaultTipViewTheme
  },
  tagView: { ...DefaultTagViewTheme
  },
  colors: { ...DefaultColorsTheme
  },
  icons: {},
  fonts: {},
  sizes: { ...DefaultSizesTheme
  }
};
exports.DefaultTheme = DefaultTheme;
