// expo-nexen-player/lib/commonjs/components/ErrorView.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _reactNative = require("react-native");

var _icons = require("../assets/icons");

var _ModalView = _interopRequireDefault(require("./ModalView"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const ErrorView = props => {
  const {
    errorText,
    theme,
    style
  } = props;
  const errorTextStyle = {
    color: theme === null || theme === void 0 ? void 0 : theme.textColor,
    fontSize: theme === null || theme === void 0 ? void 0 : theme.textSize,
    fontFamily: theme === null || theme === void 0 ? void 0 : theme.font
  };
  return /*#__PURE__*/_react.default.createElement(_ModalView.default, {
    style: [style]
  }, /*#__PURE__*/_react.default.createElement(_icons.IconXCircle, {
    size: theme === null || theme === void 0 ? void 0 : theme.iconSize,
    color: theme === null || theme === void 0 ? void 0 : theme.iconColor
  }), /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    style: [styles.text, errorTextStyle]
  }, errorText));
};

var _default = ErrorView;
exports.default = _default;

const styles = _reactNative.StyleSheet.create({
  container: {
    justifyContent: 'center',
    alignItems: 'center'
  },
  text: {
    marginTop: 5
  }
});
