// expo-nexen-player/lib/commonjs/components/GradientView.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _reactNative = require("react-native");

var _reactNativeSvg = _interopRequireWildcard(require("react-native-svg"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const GradientView = props => {
  const {
    startPoint,
    endPoint,
    startColor,
    middleColor,
    endColor,
    startOpacity,
    middleOpacity,
    endOpacity,
    startOffset,
    middleOffset,
    endOffset,
    style
  } = props;

  const [dimensions, setDimensions] = _react.default.useState({
    width: 0,
    height: 0
  });

  const onLayoutChange = async e => {
    const {
      width,
      height
    } = e.nativeEvent.layout;
    const {
      width: w,
      height: h
    } = dimensions;

    if (w !== width || h !== height) {
      setDimensions({
        width,
        height
      });
    }
  };

  return /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [styles.container, style],
    onLayout: onLayoutChange
  }, /*#__PURE__*/_react.default.createElement(_reactNativeSvg.default, {
    width: dimensions.width,
    height: dimensions.height
  }, /*#__PURE__*/_react.default.createElement(_reactNativeSvg.Defs, null, /*#__PURE__*/_react.default.createElement(_reactNativeSvg.LinearGradient, {
    id: "gradient",
    x1: startPoint === null || startPoint === void 0 ? void 0 : startPoint.x,
    y1: startPoint === null || startPoint === void 0 ? void 0 : startPoint.y,
    x2: endPoint === null || endPoint === void 0 ? void 0 : endPoint.x,
    y2: endPoint === null || endPoint === void 0 ? void 0 : endPoint.y
  }, /*#__PURE__*/_react.default.createElement(_reactNativeSvg.Stop, {
    offset: startOffset,
    stopColor: startColor,
    stopOpacity: startOpacity
  }), /*#__PURE__*/_react.default.createElement(_reactNativeSvg.Stop, {
    offset: middleOffset,
    stopColor: middleColor,
    stopOpacity: middleOpacity
  }), /*#__PURE__*/_react.default.createElement(_reactNativeSvg.Stop, {
    offset: endOffset,
    stopColor: endColor,
    stopOpacity: endOpacity
  }))), /*#__PURE__*/_react.default.createElement(_reactNativeSvg.Rect, {
    x: "0",
    y: "0",
    width: dimensions.width,
    height: dimensions.height,
    fill: "url(#gradient)"
  })));
};

var _default = GradientView;
exports.default = _default;
GradientView.defaultProps = {
  startPoint: {
    x: 0,
    y: 0
  },
  endPoint: {
    x: 0,
    y: 1
  },
  startColor: 'black',
  middleColor: 'black',
  endColor: 'black',
  startOpacity: 1,
  middleOpacity: 1,
  endOpacity: 1,
  startOffset: 0,
  middleOffset: 0.5,
  endOffset: 1
};

const styles = _reactNative.StyleSheet.create({
  container: {
    width: '100%',
    height: '25%'
  }
});
