// expo-nexen-player/lib/commonjs/components/HeaderControl.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _reactNative = require("react-native");

var _icons = require("../assets/icons");

var _IconButton = _interopRequireDefault(require("./IconButton"));

var _GradientView = _interopRequireDefault(require("./GradientView"));

var _IconTagView = _interopRequireDefault(require("./IconTagView"));

var _ColorUtil = require("../utils/ColorUtil");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

const HeaderControl = /*#__PURE__*/_react.default.forwardRef((props, ref) => {
  var _nexenTheme$sizes, _nexenTheme$colors, _nexenTheme$sizes2, _nexenTheme$colors2, _nexenTheme$sizes3, _nexenTheme$sizes4, _nexenTheme$sizes5, _nexenTheme$sizes6, _nexenTheme$sizes7, _nexenTheme$fonts;

  const {
    title,
    opacity,
    marginTop,
    fullScreen,
    locked,
    insets,
    playerConfig,
    nexenTheme,
    onBackPress,
    onAspectRatioPress,
    onMorePress
  } = props;

  const iconTagViewRef = _react.default.useRef(null);

  const ICON_SIZE = nexenTheme === null || nexenTheme === void 0 ? void 0 : (_nexenTheme$sizes = nexenTheme.sizes) === null || _nexenTheme$sizes === void 0 ? void 0 : _nexenTheme$sizes.primaryIconSize;
  const ICON_COLOR = nexenTheme === null || nexenTheme === void 0 ? void 0 : (_nexenTheme$colors = nexenTheme.colors) === null || _nexenTheme$colors === void 0 ? void 0 : _nexenTheme$colors.primaryIconColor;
  const TITLE_TEXT_SIZE = nexenTheme === null || nexenTheme === void 0 ? void 0 : (_nexenTheme$sizes2 = nexenTheme.sizes) === null || _nexenTheme$sizes2 === void 0 ? void 0 : _nexenTheme$sizes2.primaryTextSize;
  const TITLE_TEXT_COLOR = nexenTheme === null || nexenTheme === void 0 ? void 0 : (_nexenTheme$colors2 = nexenTheme.colors) === null || _nexenTheme$colors2 === void 0 ? void 0 : _nexenTheme$colors2.primaryTextColor;
  const CONTAINER_VERTICAL_PADDING = fullScreen ? (insets === null || insets === void 0 ? void 0 : insets.top) > 0 ? insets === null || insets === void 0 ? void 0 : insets.top : nexenTheme === null || nexenTheme === void 0 ? void 0 : (_nexenTheme$sizes3 = nexenTheme.sizes) === null || _nexenTheme$sizes3 === void 0 ? void 0 : _nexenTheme$sizes3.paddingVertical : nexenTheme === null || nexenTheme === void 0 ? void 0 : (_nexenTheme$sizes4 = nexenTheme.sizes) === null || _nexenTheme$sizes4 === void 0 ? void 0 : _nexenTheme$sizes4.paddingVertical;
  const CONTAINER_HORIZONTAL_PADDING = fullScreen ? ((insets === null || insets === void 0 ? void 0 : insets.left) + (insets === null || insets === void 0 ? void 0 : insets.right)) / 2 > 0 ? ((insets === null || insets === void 0 ? void 0 : insets.left) + (insets === null || insets === void 0 ? void 0 : insets.right)) / 2 : nexenTheme === null || nexenTheme === void 0 ? void 0 : (_nexenTheme$sizes5 = nexenTheme.sizes) === null || _nexenTheme$sizes5 === void 0 ? void 0 : _nexenTheme$sizes5.paddingHorizontal : nexenTheme === null || nexenTheme === void 0 ? void 0 : (_nexenTheme$sizes6 = nexenTheme.sizes) === null || _nexenTheme$sizes6 === void 0 ? void 0 : _nexenTheme$sizes6.paddingHorizontal;
  const CONTAINER_HEIGHT = (nexenTheme === null || nexenTheme === void 0 ? void 0 : (_nexenTheme$sizes7 = nexenTheme.sizes) === null || _nexenTheme$sizes7 === void 0 ? void 0 : _nexenTheme$sizes7.primaryIconSize) + 10 * 2 + 16 + CONTAINER_VERTICAL_PADDING;
  const isRTL = _reactNative.I18nManager.isRTL;
  (0, _react.useImperativeHandle)(ref, () => ({
    updateIconTagView: newState => {
      var _iconTagViewRef$curre;

      (_iconTagViewRef$curre = iconTagViewRef.current) === null || _iconTagViewRef$curre === void 0 ? void 0 : _iconTagViewRef$curre.updateState(newState);
    }
  }));

  const iconTagViewTheme = _react.default.useMemo(() => {
    var _nexenTheme$tagView, _nexenTheme$tagView2, _nexenTheme$tagView3, _nexenTheme$tagView4, _nexenTheme$tagView5, _nexenTheme$tagView6, _nexenTheme$tagView7, _nexenTheme$colors3, _nexenTheme$tagView8, _nexenTheme$colors4, _nexenTheme$tagView9, _nexenTheme$colors5, _nexenTheme$tagView10, _nexenTheme$colors6, _nexenTheme$tagView11, _nexenTheme$colors7, _nexenTheme$tagView12, _nexenTheme$colors8;

    return {
      height: nexenTheme === null || nexenTheme === void 0 ? void 0 : (_nexenTheme$tagView = nexenTheme.tagView) === null || _nexenTheme$tagView === void 0 ? void 0 : _nexenTheme$tagView.height,
      cornerRadius: nexenTheme === null || nexenTheme === void 0 ? void 0 : (_nexenTheme$tagView2 = nexenTheme.tagView) === null || _nexenTheme$tagView2 === void 0 ? void 0 : _nexenTheme$tagView2.cornerRadius,
      borderWidth: nexenTheme === null || nexenTheme === void 0 ? void 0 : (_nexenTheme$tagView3 = nexenTheme.tagView) === null || _nexenTheme$tagView3 === void 0 ? void 0 : _nexenTheme$tagView3.borderWidth,
      iconSize: nexenTheme === null || nexenTheme === void 0 ? void 0 : (_nexenTheme$tagView4 = nexenTheme.tagView) === null || _nexenTheme$tagView4 === void 0 ? void 0 : _nexenTheme$tagView4.iconSize,
      textSize: nexenTheme === null || nexenTheme === void 0 ? void 0 : (_nexenTheme$tagView5 = nexenTheme.tagView) === null || _nexenTheme$tagView5 === void 0 ? void 0 : _nexenTheme$tagView5.textSize,
      textLineHeight: nexenTheme === null || nexenTheme === void 0 ? void 0 : (_nexenTheme$tagView6 = nexenTheme.tagView) === null || _nexenTheme$tagView6 === void 0 ? void 0 : _nexenTheme$tagView6.textLineHeight,
      iconColor: (nexenTheme === null || nexenTheme === void 0 ? void 0 : (_nexenTheme$tagView7 = nexenTheme.tagView) === null || _nexenTheme$tagView7 === void 0 ? void 0 : _nexenTheme$tagView7.iconColor) || (nexenTheme === null || nexenTheme === void 0 ? void 0 : (_nexenTheme$colors3 = nexenTheme.colors) === null || _nexenTheme$colors3 === void 0 ? void 0 : _nexenTheme$colors3.primaryColor),
      activeIconColor: (nexenTheme === null || nexenTheme === void 0 ? void 0 : (_nexenTheme$tagView8 = nexenTheme.tagView) === null || _nexenTheme$tagView8 === void 0 ? void 0 : _nexenTheme$tagView8.activeIconColor) || (0, _ColorUtil.getAlphaColor)(nexenTheme === null || nexenTheme === void 0 ? void 0 : (_nexenTheme$colors4 = nexenTheme.colors) === null || _nexenTheme$colors4 === void 0 ? void 0 : _nexenTheme$colors4.accentColor, 0.7),
      inactiveIconColor: (nexenTheme === null || nexenTheme === void 0 ? void 0 : (_nexenTheme$tagView9 = nexenTheme.tagView) === null || _nexenTheme$tagView9 === void 0 ? void 0 : _nexenTheme$tagView9.inactiveIconColor) || (0, _ColorUtil.getAlphaColor)(nexenTheme === null || nexenTheme === void 0 ? void 0 : (_nexenTheme$colors5 = nexenTheme.colors) === null || _nexenTheme$colors5 === void 0 ? void 0 : _nexenTheme$colors5.primaryColor, 0.5),
      textColor: (nexenTheme === null || nexenTheme === void 0 ? void 0 : (_nexenTheme$tagView10 = nexenTheme.tagView) === null || _nexenTheme$tagView10 === void 0 ? void 0 : _nexenTheme$tagView10.textColor) || (nexenTheme === null || nexenTheme === void 0 ? void 0 : (_nexenTheme$colors6 = nexenTheme.colors) === null || _nexenTheme$colors6 === void 0 ? void 0 : _nexenTheme$colors6.secondaryTextColor),
      borderColor: (nexenTheme === null || nexenTheme === void 0 ? void 0 : (_nexenTheme$tagView11 = nexenTheme.tagView) === null || _nexenTheme$tagView11 === void 0 ? void 0 : _nexenTheme$tagView11.borderColor) || (0, _ColorUtil.getAlphaColor)(nexenTheme === null || nexenTheme === void 0 ? void 0 : (_nexenTheme$colors7 = nexenTheme.colors) === null || _nexenTheme$colors7 === void 0 ? void 0 : _nexenTheme$colors7.accentColor, 0.7),
      backgroundColor: (nexenTheme === null || nexenTheme === void 0 ? void 0 : (_nexenTheme$tagView12 = nexenTheme.tagView) === null || _nexenTheme$tagView12 === void 0 ? void 0 : _nexenTheme$tagView12.backgroundColor) || (0, _ColorUtil.getAlphaColor)(nexenTheme === null || nexenTheme === void 0 ? void 0 : (_nexenTheme$colors8 = nexenTheme.colors) === null || _nexenTheme$colors8 === void 0 ? void 0 : _nexenTheme$colors8.primaryColor, 0.0)
    };
  }, [nexenTheme, fullScreen]);

  const titleTextStyle = {
    fontSize: TITLE_TEXT_SIZE,
    color: TITLE_TEXT_COLOR,
    fontFamily: nexenTheme === null || nexenTheme === void 0 ? void 0 : (_nexenTheme$fonts = nexenTheme.fonts) === null || _nexenTheme$fonts === void 0 ? void 0 : _nexenTheme$fonts.primaryFont
  };
  return /*#__PURE__*/_react.default.createElement(_reactNative.Animated.View, {
    style: [styles.container, {
      height: CONTAINER_HEIGHT
    }, {
      opacity,
      marginTop
    }]
  }, /*#__PURE__*/_react.default.createElement(_GradientView.default, {
    style: {
      width: '100%',
      height: '100%'
    },
    startOpacity: 0.5,
    middleOpacity: 0.2,
    endOpacity: 0.0
  }), /*#__PURE__*/_react.default.createElement(_reactNative.TouchableHighlight, {
    style: [styles.innerContainer, {
      paddingTop: CONTAINER_VERTICAL_PADDING,
      paddingHorizontal: CONTAINER_HORIZONTAL_PADDING
    }]
  }, /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: {
      flex: 1
    }
  }, /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: {
      flexDirection: 'row',
      justifyContent: 'space-between'
    }
  }, /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [styles.iconButtonContainer, {
      flexShrink: 1
    }]
  }, !locked && !(playerConfig !== null && playerConfig !== void 0 && playerConfig.disableBack) && /*#__PURE__*/_react.default.createElement(_IconButton.default, {
    style: {
      transform: [{
        scaleX: isRTL ? -1 : 1
      }]
    },
    onPress: onBackPress
  }, /*#__PURE__*/_react.default.createElement(_icons.IconChevronLeft, {
    size: ICON_SIZE,
    color: ICON_COLOR
  })), /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    style: [styles.titleText, titleTextStyle],
    numberOfLines: 2
  }, title === null || title === void 0 ? void 0 : title.toUpperCase())), /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [styles.iconButtonContainer]
  }, !locked && (playerConfig === null || playerConfig === void 0 ? void 0 : playerConfig.layoutMode) === 'advanced' && /*#__PURE__*/_react.default.createElement(_IconTagView.default, {
    ref: iconTagViewRef,
    style: {
      marginHorizontal: 4
    },
    theme: iconTagViewTheme
  }), !locked && !(playerConfig !== null && playerConfig !== void 0 && playerConfig.disableResizeMode) && (playerConfig === null || playerConfig === void 0 ? void 0 : playerConfig.layoutMode) === 'intermediate' && /*#__PURE__*/_react.default.createElement(_IconButton.default, {
    onPress: onAspectRatioPress
  }, /*#__PURE__*/_react.default.createElement(_icons.IconAspectRatio, {
    size: ICON_SIZE,
    color: ICON_COLOR
  })), !locked && !(playerConfig !== null && playerConfig !== void 0 && playerConfig.disableMore) && (playerConfig === null || playerConfig === void 0 ? void 0 : playerConfig.layoutMode) === 'advanced' && /*#__PURE__*/_react.default.createElement(_IconButton.default, {
    onPress: onMorePress
  }, /*#__PURE__*/_react.default.createElement(_icons.IconMoreHorizontal, {
    size: ICON_SIZE,
    color: ICON_COLOR
  })))))));
});

var _default = /*#__PURE__*/_react.default.memo(HeaderControl);

exports.default = _default;
HeaderControl.defaultProps = {};

const styles = _reactNative.StyleSheet.create({
  container: {
    position: 'absolute',
    left: 0,
    right: 0,
    top: 0,
    height: 56,
    minHeight: 56,
    zIndex: 100
  },
  innerContainer: {
    position: 'absolute',
    left: 0,
    top: 0,
    right: 0,
    bottom: 0
  },
  iconButtonContainer: {
    flexDirection: 'row',
    alignItems: 'center'
  },
  textContainer: {},
  titleText: {
    fontSize: 20,
    paddingHorizontal: 10,
    flexShrink: 1
  }
});
