// expo-nexen-player/lib/commonjs/components/LineSeekBar.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _reactNative = require("react-native");

var _MathUtil = require("../utils/MathUtil");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const LineSeekBar = props => {
  const {
    trackTime,
    totalTrackTime,
    layoutWidth,
    theme
  } = props;
  const isRTL = _reactNative.I18nManager.isRTL;
  const rtlMultiplier = isRTL ? -1 : 1;

  const seekWidth = _react.default.useRef(new _reactNative.Animated.Value(0)).current;

  const seekableDistance = _react.default.useRef(0);

  _react.default.useEffect(() => {
    seekableDistance.current = layoutWidth * rtlMultiplier;
    const position = (0, _MathUtil.boundPosition)((0, _MathUtil.trackToSeekValue)(trackTime, totalTrackTime, seekableDistance.current), seekableDistance.current);
    seekWidth.setValue(position * rtlMultiplier);
  }, [trackTime, totalTrackTime, layoutWidth]);

  return /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: styles.container
  }, /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [styles.totalTrack, {
      height: theme === null || theme === void 0 ? void 0 : theme.lineHeight,
      borderRadius: (theme === null || theme === void 0 ? void 0 : theme.lineHeight) / 2,
      backgroundColor: theme === null || theme === void 0 ? void 0 : theme.lineUnderlayColor
    }]
  }), /*#__PURE__*/_react.default.createElement(_reactNative.Animated.View, {
    style: [styles.track, {
      width: seekWidth,
      height: theme === null || theme === void 0 ? void 0 : theme.lineHeight,
      borderRadius: (theme === null || theme === void 0 ? void 0 : theme.lineHeight) / 2,
      backgroundColor: theme === null || theme === void 0 ? void 0 : theme.lineColor
    }]
  }));
};

var _default = LineSeekBar;
exports.default = _default;

const styles = _reactNative.StyleSheet.create({
  container: {
    position: 'absolute',
    left: 0,
    right: 0,
    bottom: 0,
    justifyContent: 'center',
    zIndex: 999
  },
  totalTrack: {
    position: 'absolute',
    width: '100%',
    height: 2,
    borderRadius: 1
  },
  track: {
    height: 2,
    borderRadius: 1
  }
});
