// expo-nexen-player/lib/commonjs/components/Loader.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _reactNative = require("react-native");

var _icons = require("../assets/icons");

var _ModalView = _interopRequireDefault(require("./ModalView"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const Loader = props => {
  const {
    theme,
    style
  } = props;

  const spinValue = _react.default.useRef(new _reactNative.Animated.Value(0)).current;

  const spinAnim = spinValue.interpolate({
    inputRange: [0, 1],
    outputRange: ['0deg', '360deg']
  });

  _react.default.useEffect(() => {
    const anim = _reactNative.Animated.loop(_reactNative.Animated.timing(spinValue, {
      toValue: 1,
      duration: 1600,
      easing: _reactNative.Easing.linear,
      useNativeDriver: true
    }));

    anim.start();
    return () => anim.stop();
  }, []);

  return /*#__PURE__*/_react.default.createElement(_ModalView.default, {
    style: style
  }, /*#__PURE__*/_react.default.createElement(_reactNative.Animated.View, {
    style: {
      transform: [{
        rotate: spinAnim
      }]
    }
  }, props.loaderIcon ? props.loaderIcon : /*#__PURE__*/_react.default.createElement(_icons.IconLoader, {
    size: theme === null || theme === void 0 ? void 0 : theme.iconSize,
    color: theme === null || theme === void 0 ? void 0 : theme.iconColor
  })), /*#__PURE__*/_react.default.createElement(_reactNative.View, null, /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    style: [styles.text, {
      color: theme === null || theme === void 0 ? void 0 : theme.textColor,
      fontSize: theme === null || theme === void 0 ? void 0 : theme.textSize
    }]
  }, props.loaderText)));
};

var _default = /*#__PURE__*/_react.default.memo(Loader);

exports.default = _default;

const styles = _reactNative.StyleSheet.create({
  container: {
    position: 'absolute',
    justifyContent: 'center',
    alignItems: 'center',
    paddingHorizontal: 16,
    paddingVertical: 12
  },
  text: {
    marginTop: 5
  }
});
