// expo-nexen-player/lib/commonjs/components/PlayButton.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _reactNative = require("react-native");

var _react = _interopRequireDefault(require("react"));

var _icons = require("../assets/icons");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const PlayButton = props => {
  const {
    dimension
  } = props;
  const minValue = Math.min(dimension === null || dimension === void 0 ? void 0 : dimension.width, dimension === null || dimension === void 0 ? void 0 : dimension.height);
  const ICON_SIZE = minValue * 0.3;
  return /*#__PURE__*/_react.default.createElement(_reactNative.TouchableOpacity, {
    style: styles.container,
    activeOpacity: 0.6,
    onPress: props.onPlayPress
  }, /*#__PURE__*/_react.default.createElement(_icons.IconPlayCircle, {
    size: ICON_SIZE,
    color: "rgba(255,255,255,0.7)"
  }));
};

var _default = PlayButton;
exports.default = _default;
PlayButton.defaultProps = {
  dimension: {
    width: 0,
    height: 0
  }
};

const styles = _reactNative.StyleSheet.create({
  container: {
    position: 'absolute',
    width: 72,
    height: 72,
    justifyContent: 'center',
    alignItems: 'center'
  }
});
