// expo-nexen-player/lib/commonjs/components/PosterView.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _reactNative = require("react-native");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const PosterView = props => {
  const {
    posterSource,
    posterResizeMode
  } = props;
  const imageStyle = {
    resizeMode: posterResizeMode
  };
  return /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: styles.container
  }, /*#__PURE__*/_react.default.createElement(_reactNative.Image, {
    style: [styles.image, imageStyle],
    source: {
      uri: posterSource
    }
  }));
};

var _default = PosterView;
exports.default = _default;

const styles = _reactNative.StyleSheet.create({
  container: {
    position: 'absolute',
    left: 0,
    top: 0,
    right: 0,
    bottom: 0,
    backgroundColor: '#0a0a0a'
  },
  image: {
    width: '100%',
    height: '100%'
  }
});
