// expo-nexen-player/lib/commonjs/components/TimeTagView.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _reactNative = require("react-native");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const TimeTagView = props => {
  const {
    timeText,
    theme,
    style
  } = props;
  const textContainerStyle = {
    height: theme === null || theme === void 0 ? void 0 : theme.height,
    borderRadius: theme === null || theme === void 0 ? void 0 : theme.cornerRadius,
    borderWidth: theme === null || theme === void 0 ? void 0 : theme.borderWidth,
    borderColor: theme === null || theme === void 0 ? void 0 : theme.borderColor,
    backgroundColor: theme === null || theme === void 0 ? void 0 : theme.backgroundColor
  };
  const timeTextStyle = {
    fontSize: theme === null || theme === void 0 ? void 0 : theme.textSize,
    lineHeight: theme === null || theme === void 0 ? void 0 : theme.textLineHeight,
    color: theme === null || theme === void 0 ? void 0 : theme.textColor,
    fontFamily: theme === null || theme === void 0 ? void 0 : theme.font
  };
  return /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [styles.container, style, textContainerStyle]
  }, /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    style: [styles.text, timeTextStyle]
  }, timeText));
};

var _default = TimeTagView;
exports.default = _default;

const styles = _reactNative.StyleSheet.create({
  container: {
    minWidth: 40,
    height: 18,
    borderRadius: 9,
    borderWidth: 1,
    justifyContent: 'center',
    alignItems: 'center'
  },
  text: {
    fontSize: 11,
    marginHorizontal: 6,
    lineHeight: 13,
    opacity: 0.8
  }
});
