// expo-nexen-player/lib/module/utils/Theme.js
export const DefaultColorsTheme = {
  primaryColor: '#0a0a0a',
  secondaryColor: '#fafafa',
  accentColor: '#fa5005',
  primaryTextColor: '#fafafa',
  secondaryTextColor: '#f2f2f2',
  primaryIconColor: '#fafafa',
  secondaryIconColor: '#f2f2f2',
  rippleIconColor: '#f2f2f2'
};
export const DefaultSizesTheme = {
  modalCornerRadius: 12,
  primaryTextSize: 20,
  secondaryTextSize: 15,
  primaryIconSize: 20,
  secondaryIconSize: 20,
  rippleIconSize: 28,
  paddingVertical: 8,
  paddingHorizontal: 4
};
export const DefaultTrackSeekBarTheme = {
  trackHeight: 2,
  thumbSize: 10,
  thumbBorderWidth: 5,
  thumbCornerRadius: 5,
  thumbBorderCornerRadius: 7.5
};
export const DefaultSpeedSeekBar = {
  textSize: 12,
  lineHeight: 2,
  dotSize: 6,
  dotCornerRadius: 3,
  thumbSize: 18,
  thumbCornerRadius: 9
};
export const DefaultVolumeSeekBarTheme = {
  barColor: 'rgba(250,250,250,0.5)',
  underlayColor: 'rgba(10,10,10,0.3)'
};
export const DefaultBrightnessSeekBarTheme = {
  barColor: 'rgba(250,250,250,0.5)',
  underlayColor: 'rgba(10,10,10,0.3)'
};
export const DefaultMiniSeekBarTheme = {
  barColor: 'rgba(250,80,5,0.6)',
  underlayColor: 'rgba(250,250,250,0.3)',
  thumbColor: '#fafafa',
  trackHeight: 2,
  thumbSize: 10,
  thumbCornerRadius: 5
};
;
export const DefaultLockButtonTheme = {
  textSize: 14,
  thumbIconSize: 18,
  thumbHeight: 36,
  thumbWidth: 36,
  thumbCornerRadius: 18
};
export const DefaultTagViewTheme = {
  height: 18,
  borderWidth: 1,
  cornerRadius: 9,
  textSize: 12,
  textLineHeight: 14,
  iconSize: 12
};
export const DefaultTipViewTheme = {
  iconSize: 20,
  textSize: 18
};
export const DefaultTheme = {
  trackSeekBar: { ...DefaultTrackSeekBarTheme
  },
  speedSeekBar: { ...DefaultSpeedSeekBar
  },
  miniSeekBar: { ...DefaultMiniSeekBarTheme
  },
  lockButton: { ...DefaultLockButtonTheme
  },
  tipView: { ...DefaultTipViewTheme
  },
  tagView: { ...DefaultTagViewTheme
  },
  colors: { ...DefaultColorsTheme
  },
  icons: {},
  fonts: {},
  sizes: { ...DefaultSizesTheme
  }
};
